src/core/device.c

```diff
@@ -114,12 +159,134 @@ static void device_done(Unit *u) {
+                         minor(dev),
+                         io_cost_qos->enabled,
+                         io_cost_ctrl_to_string(io_cost_qos->ctrl),
+                         io_cost_qos->read_latency_percentile / 100,
+                         io_cost_qos->read_latency_percentile % 100,
+                         io_cost_qos->read_latency_threshold,
+                         io_cost_qos->write_latency_percentile / 100,
+                         io_cost_qos->write_latency_percentile % 100,
+                         io_cost_qos->write_latency_threshold,
+                         io_cost_qos->min / 100,
+                         io_cost_qos->min % 100,
+                         io_cost_qos->max / 100,
+                         io_cost_qos->max % 100);
+
+        } else
+                return 0;
+
+        (void) set_attribute_and_warn(root_slice_unit, "io", "io.cost.qos", buf);
+
+        return 0;
+}
+
+static int root_cgroup_apply_block_device_io_cost_model(Unit *root_slice_unit, Device *d) {
+        char buf[DECIMAL_STR_MAX(dev_t)*2+9+12+(9+DECIMAL_STR_MAX(uint64_t)+1)*6];
+        IOCostModel *io_cost_model;
+        dev_t dev;
+        int r;
+
+        assert(root_slice_unit);
+        assert(d);
+
+        io_cost_model = &d->io_cost_model;
+
+        /* If ctrl is invalid it means that io_cost_model is not configured
+         * for this device */
+        if (!io_cost_model || io_cost_model->ctrl == _IO_COST_CTRL_INVALID)
+                return 0;
+
+        if (!verify_io_cost_supported(d))
+                return log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "io cost is not supported");
+
+        r = lookup_block_device(d->devname, &dev);
+        if (r < 0)
+                return r;
+
+        if (io_cost_model->ctrl == IO_COST_CTRL_AUTO) {
+                log_info("ctrl is set to auto for io_cost_model, setting all other values to kernel default");
+                xsprintf(buf, "%u:%u ctrl=%.4s\n",
+                         major(dev),
+                         minor(dev),
+                         io_cost_ctrl_to_string(IO_COST_CTRL_AUTO));
+
+        } else if (io_cost_model->ctrl == IO_COST_CTRL_USER) {
+                xsprintf(buf, "%u:%u ctrl=%.4s model=linear rbps=%" PRIu64 " rseqiops=%" PRIu64 " rrandiops=%" PRIu64 " wbps=%" PRIu64 " wseqiops=%" PRIu64 " wrandiops=%" PRIu64 "\n",
+                         major(dev),
+                         minor(dev),
+                         io_cost_ctrl_to_string(io_cost_model->ctrl),
+                         io_cost_model->rbps,
+                         io_cost_model->rseqiops,
+                         io_cost_model->rrandiops,
+                         io_cost_model->wbps,
+                         io_cost_model->wseqiops,
+                         io_cost_model->wrandiops);
+
+        } else
+                return 0;
+
+        (void) set_attribute_and_warn(root_slice_unit, "io", "io.cost.model", buf);
+
+        return 0;
+}
+
 static int device_load(Unit *u) {
```

Comment:
Would the assignment work in the `device_load()`? The unit may be constructed in the systemd alredy but kernel will only discover it (possibly) later, then uevent, then udev, then the unit activates in systemd (i.e. config should be applied after unit state change).
At the same time, another call is also needed when the property is updated dynamically on an existing device.

Reply:
The update will happen after `systemctl daemon-reload` is run. The intention is for the user to be able to specify these fields via overrides in .device.d/ directories. The daemon reload would be required. There would be no other way to specify the io.cost parameters for a device, but that is the intention of this change. The purpose of adding this code to the device_load would be to do the assignment and write to /sys/fs/cgroup/io.cost.*

Is there an alternative approach you would suggest?

Reply:
Device units are not actively controlled by systemd. systemd only creates wrapping jobs and waits passively for a notification when the device pops up, then the unit is activated (it's loaded earlier).
You can have the .device.d config file, you can call daemon-reload but if the device doesn't exist yet (e.g. USB stick or during boot) the io.cost parameters won't be applied (ENODEV).

I see that [a load may be triggered](https://github.com/systemd/systemd/blob/bb916f350a090bf10a8e71ea26147d4071300d2a/src/core/device.c#L562) in response to an [udev notification](https://github.com/systemd/systemd/blob/bb916f350a090bf10a8e71ea26147d4071300d2a/src/core/device.c#L979) but I'm afraid it won't work that way if the unit was "loaded" -- that's why I asked whether this works.

The possible spot to hook into seems `device_update_found_by_sysfs()` (going upwards from `device_found_changed()`).